draft-ietf-tsvwg-careful-resume.xml

```diff
@@ -636,6 +659,10 @@ Observing ...> Connect -> Reconnaissance --------------------> Normal
             The value of ssthresh on leaving the Safe Retreat Phase
             MUST NOT be more than the PipeSize.</t>
         </list></t>
+
+        <t>The safe retreat process is not used when the connection uses BBR. Instead,
+           if the remembered values were excessive, BBR will naturally enter its Drain
+           state. </t>
      
```

Comment:
To prevent suppressing other flows, it is important that the rate used in Safe Retreat is not directly derived from the packets in flight during a failed unvalidated period. This is because the unvalidated phase - in the worst case - can send at a rate significantly more than a reasonable share of the capacity (e.g. after a path change, or mnay flows have started since the observed rate). Draining needs to be very conservative (balancing a potentially very aggressive jump) to take account that an excessive unvalidated rate could have (unfairly) pushed-out other flows from a congested bottleneck. 
Your current suggetion that we say BBR Drain is equivalent, might be true depending on how this implemented, but it makes me unsure: I'm therefore hessitant about this proposed text. I'd be happier if we said "BBR needs to use a  method that is equivalent to Safe Retreat".

Reply:
During drain, the pacing rate is specified as:
~~~
  BBREnterDrain():
    BBR.state = Drain
    BBR.pacing_gain = 1/BBRStartupCwndGain  /* pace slowly */
    BBR.cwnd_gain = BBRStartupCwndGain      /* maintain cwnd */
~~~
The pacing gain is applied to the estimated bottleneck bandwidth, which results from measurements -- not to the probing rate remembered from the previous connection. BBR, the estimated  bottleneck bandwidth will be set to the capacity of the bottleneck, or below if there is competing traffic -- the measured bandwidth can never exceed the bandwidth of the path. With the formula above, the pacing rate will be set to about 36% of the measured bandwidth, and the CWND to 2*minRTT*measured bandwidth -- i.e., at most 2*BDP. The connection will send no packet at all until the bytes in flight drop below 2*BDP, then will continue at 36% of actual bandwidth until the queue is drained.

I think that this meets the "don't starve others" criteria set in RFC5033bis. We may debate whether it results in adequate fairness, but this is a general issue with BBR, not a specific issue with careful resume. The "36% of bottleneck bandwidth" also happens without careful resume, as a normal consequence of exponential startup.

Reply:
Thanks, I agree with your description of how BBR ought to work, that was useful.
* However, my issue is more subtle. Consider the case where many flows now use the capacity after it was once measured by the observation. When the sender makes the CR unvalidated jump (rather than normal exponential increase), the unvalidated phase would result in a BBR flow measuring the bottleneck bandwidth at a time when that flow could be very significantly overloading the bottleneck it is measuring (the overload could be much more than twice the previous load it contributed, as allowed with exponential growth). 
* Although the measured bandwidth cannot exceed the actual bottleneck limit, I think the limit of 36% of the measured capacity can in this specific case result in a very unreasonable share of the capacity.  
* e.g., a flow could claim an arbitrary large share of the capacity depending on how large a jump, rather than being bounded to twice the previous probed rate. If a flow did that, it would then suppress all other flows. This motivates being even morer careful, and not directly using that measured rate, or the suggestion to at least slow-starting towards that measured rate.

Reply:
Yes, temporary excess could result in higher share. That's the fairness problem. It is also a problem with startup in general. Suppose N established flows, and flow N+1 engages in exponential growth. It will keep pushing until "the measured bandwidth does not increase any more", arguably with the same final measured bandwidth as what we get with "(not so) careful resume". The not so careful version just gets there sooner.

I think the solution to that problem is to have a long term equalizer, so the excess bandwidth share is eventually reduced. That, or running an AQM at the bottleneck...

Reply:
Also, you are making the hypothesis that the other flows sharing the bottleneck will reduce their sending rate. I am not sure that they have the time to do that. Let's look at the time line:

* T, start sending at the "resume rate",
* T + 1 RTT, packet sent at time T is acked, end of round 1,
* T + X RTT, last packet sent in round 1 is acknowledged, end of round 2.

Depending on queues, etc., the second round lasts between 1 and 2 or more RTT. Typically 2 if buffers at the bottleneck are sized as 2*BDP, but it could be more in case of buffer bloat. Competing traffic will notice the competition somewhere during round 2. That does not leave much time to react and reduce bandwidth, so we may expect that during round 2 a good fraction of the bottleneck bandwidth will be used by competing flow. The measurement at the end of round 2 will reflect that competition.

The excessive bandwidth, if it is excessive, will cause a packet loss that will be noticed by the end of round 2. BBR will enter Drain at that point, and start sending at a very low rate. Other flows will also notice the packet loss, and will reduce their data rate -- but by the time they reduce their rate, the measurement is done and BBR is not sending at the excess rate any more.

I don't think that we have much of a problem.

Reply:
The updated text mitigates the issue by ignoring the measurements received during Drain. The pacing rate during Drain will be set to 36% of the latest measurement received before a loss (or maybe delay increase) is observed. I think that this addresses the concern.